### PR TITLE
Add nonstd/optional.hpp library

### PR DESCRIPTION
**Pull request by LAJW**
This will allow us to unify and enforce proper error handling while at the same time avoid exceptions.

I'd like to include it as a submodule but @peterschrammel suggested I should just drop it in as a single file.

All credit goes to: https://github.com/martinmoene/optional-lite

### CODE REVIEW

**Comment by LAJW** [comment]
@reuk @tautschnig @smowton You might want to chip in as well.

**Comment by reuk** [comment]
Seems good - not sure what else there is to say.

**Comment by tautschnig** [comment]
@LAJW Would you mind implementing some first use case as well? I can see that it is a _nice_ language feature, but I would like to understand better how to make it a _useful_ language feature in the context of this code base. Thanks!

**Comment by reuk** [comment]
I thought of something to say: the code itself is obviously useful. Having stack-bound nullable objects is invaluable.

What's not so great is including the file directly into the repository. The number of external dependencies that are being added to this project are steadily growing (miniz, minisat, catch, this file...), and there are better approaches to dependency management than directly including the dependencies in the source tree. I recommend using [CMake](https://cmake.org/) in conjunction with [Hunter](https://github.com/ruslo/hunter).

**Comment by LAJW** [comment]
@tautschnig I don't want to include a random change (as there are too many potential candidates).

Here's a clear example showing why it's superior to what we have now:

The old way:
```c++
exprt expr; // Can't be made const even if it's used in a const way
if (some_function(expr)) // Easy to forget the check, also did the function succeed or fail here?
{
   // do stuff with expr
}
```

The new way:
```c++
const optional<exprt> maybe_expr=some_function();
if (maybe_expr.has_value())
{
  const expr=maybe_expr.value(); // Throws if there's no value. Enforces error checking.
  // do stuff with expr
}
```

As you can see the problem with returning boolean to signal the error is that it's easy to forget whether false or true means success or the other way around and easy to use invalid object and the object can't be made const.

And if you don't like verbosity, you can use operators:

```c++
const optional<exprt> maybe_expr=some_function();
if (maybe_expr)
{
  const expr=*maybe_expr;
  // do stuff with expr
}
```

**Comment by tautschnig** [comment]
> What's not so great is including the file directly into the repository. The number of external dependencies that are being added to this project are steadily growing (miniz, minisat, catch, this file...), and there are better approaches to dependency management than directly including the dependencies in the source tree. I recommend using CMake in conjunction with Hunter.

I do agree that including external source is not great as it has both maintenance (including security) and legal ramifications. Until recently, the number of such external components has been very limited and been growing at a very slow pace (miniz just being the second one after big-int, then catch came along). The conflict here is: what is best for the customer vs. what is best for developers. Customers need to be able to build the tool in/for their desired environment.

(I do refrain from naming particular technology or tools as solution at this point.)

**Comment by martin-cs** [comment]
On Wed, 2017-08-23 at 16:03 +0000, Michael Tautschnig wrote:
> > What's not so great is including the file directly into the repository. The number of external dependencies that are being added to this project are steadily growing (miniz, minisat, catch, this file...), and there are better approaches to dependency management than directly including the dependencies in the source tree. I recommend using CMake in conjunction with Hunter.
> 
> I do agree that including external source is not great as it has both maintenance (including security) and legal ramifications. Until recently, the number of such external components has been very limited and been growing at a very slow pace (miniz just being the second one after big-int, then catch came along). The conflict here is: what is best for the customer vs. what is best for developers. Customers need to be able to build the tool in/for their desired environment.
> 
> (I do refrain from naming particular technology or tools as solution at this point.)

Adding dependencies used to be a very frowned upon activity.  I do not
know if this is still the case.

**Comment by LAJW** [comment]
> Adding dependencies used to be a very frowned upon activity.  I do not know if this is still the case.

I believe this refers to system dependencies (requiring users to `apt-get install libz-dev` for example). Which indeed introduces a lot of problems (since we're trying to be multiplatform) and should be avoided at all costs.

Using static libraries (or header-only libraries as in this case) isn't a huge problem as long as they compile under platforms we support. We just don't have a streamlined way of handling these. Including the source is a solution, but not the ideal one, as we don't have a reference to the original repository so updating the library requires unnecessary effort. Submodules or package managers are a solution to that problem.

What isn't a solution is not including libraries whatsoever and reinventing the wheel introducing bugs along with it.

**Comment by tautschnig** [comment]
@LAJW Thank you very much for the example. How about switching to C++ 17 instead of using an external workaround?

**Comment by LAJW** [comment]
@tautschnig That would be the ideal solution. Sadly Visual Studio 2013 (AppVeyor) doesn't support it. And I'm not sure about clang/gcc since C++17 was standardized just this year.

**Comment by tautschnig** [comment]
What is the urgency? Does it justify introducing an external dependency? (I am not necessarily expecting public answers, you might just want to elaborate this with your management chain.)

**Comment by LAJW** [comment]
@tautschnig Because we're gonna be stuck with C++11 (for backward compatibility) for the foreseeable future and we don't have 6 years to wait.

**Comment by tautschnig** [comment]
> Using locally compiled static libraries (or header-only libraries as in this case) isn't a huge problem as long as they compile under platforms we support. We just don't have a streamlined way of handling these. Including the source is a solution, but not the ideal one, as we don't have a reference to the original repository so updating the library requires unnecessary effort. Submodules or package managers are a solution to that problem.

Please do keep reproducibility in mind: at present, checking out a particular version of the git repository is guaranteed to produce exactly the same results. (MiniSat being a slight caveat, but in practice it hasn't changed in years.) With any external dependencies there need to be additional measures in place to cater for this. git submodules of course would be such a means. Package managers (whether distributions or language-specific ones) do require additional care.

Another aspect to consider is: can you build from source without a network connection? This is essential in some current users' scenarios.

**Comment by peterschrammel** [comment]
Optional is optional and low priority in general, but it would allow us the remove pointer usage and various workarounds with iterators and booleans in many places. 
I suggested to create this PR to start the discussion whether we want to consider introducing it in the codebase. Putting it into the source tree is least invasive for our users. I expect that there is minimal maintenance overhead and it is easy to kick out again once we switch to C++17.

**Comment by martin-cs** [comment]
On Wed, 2017-08-23 at 16:23 +0000, Łukasz A.J. Wrona wrote:
> > Adding dependencies used to be a very frowned upon activity.  I do not know if this is still the case.
> 
> I believe this refers to system dependencies (requiring users to `apt-get install libz-dev` for example).
<snip>
> What isn't a solution is not including libraries whatsoever and reinventing the wheel introducing bugs along with it.

You misunderstand my comment.  I'm not saying about what's right or
wrong, I'm saying about what is at the moment.  The last I heard, the
policy is to not add external dependencies unless it is absolutely
non-impossible to solve an immediate user / customer need without doing
so.  I would not recommend attempting a change of policy via PR; other
people have tried and it has descended into argument and got ignored.  I
would suggest getting Daniel (project over-all), Peter (diffblue) or
Michael (most experienced / senior dev) to clarify what the policy is
and then go from there.  If you want to change the policy, make the case
to them.  However be warned that there are a significant number of users
of CPROVER outside of Diffblue and you may not be aware of their
deployment platforms, licensing restrictions, development environments,
etc.  This is one of the reasons why the project has traditionally been
quite conservative with regard to deprecating older compilers, moving to
newer language versions, etc.

HTH

File: src/nonstd/optional.hpp [hunk 1]
```diff
@@ -0,0 +1,1087 @@
...
+  : logic_error( "bad optional access" ) {}
+};
+
+/// optional
+
+template< typename T>
+class optional
+{
+private:
+    typedef void (optional::*safe_bool)() const;
```
Comment (NathanJPhillips) [inline src/nonstd/optional.hpp:578]: Inconsistent indentation (previous class used our standard of two spaces).

File: src/nonstd/optional.hpp [hunk 2]
```diff
@@ -0,0 +1,1087 @@
...
+#ifndef  optional_CONFIG_MAX_ALIGN_HACK
+# define optional_CONFIG_MAX_ALIGN_HACK  0
+#endif
+
+#ifndef  optional_CONFIG_ALIGN_AS
+// no default, used in #if defined()
+#endif
+
+#ifndef  optional_CONFIG_ALIGN_AS_FALLBACK
+# define optional_CONFIG_ALIGN_AS_FALLBACK  double
```
Comment (NathanJPhillips) [inline src/nonstd/optional.hpp:38]: Including this file leaves macros defined that are only for internal use.

File: src/nonstd/optional.hpp [hunk 3]
```diff
@@ -0,0 +1,1087 @@
...
+#endif
+
+// Compiler detection (C++17 is speculative):
+
+#define optional_CPP11_OR_GREATER  ( __cplusplus >= 201103L )
```
Comment (NathanJPhillips) [inline src/nonstd/optional.hpp:43]: We already assume C++11 in our code, doing so here could greatly simplify things.

File: src/nonstd/optional.hpp [hunk 4]
```diff
@@ -0,0 +1,1087 @@
...
+    optional_ALIGN_TYPE( float * ),
+    optional_ALIGN_TYPE( double * ),
+    optional_ALIGN_TYPE( long double * ),
+
+    optional_ALIGN_TYPE( Unknown ( * )( Unknown ) ),
+    optional_ALIGN_TYPE( Unknown * Unknown::*     ),
+    optional_ALIGN_TYPE( Unknown ( Unknown::* )( Unknown ) ),
+
+    nulltype
+    > > > > > > >    > > > > > > >
```
Comment (NathanJPhillips) [inline src/nonstd/optional.hpp:417]: This definitely needs a comment explaining it.

**Comment by LAJW** [comment]
@NathanJPhillips I'm not the author, here's the original version: https://github.com/martinmoene/optional-lite It's already tested, no need to add any of our own.

It's supposed to be a shim for the C++17 std::optional which doesn't support reference types (but can be used with std::reference_wrapper as other std::containers). It's trying to adhere to standard unlike Boost version (which in addition has several dependencies of its own).

**Comment by tautschnig** [comment]
> It's already tested, no need to add any of our own.

This level of trust on someone else's QA work seems inappropriate to me. In addition to plain QA, and as @NathanJPhillips pointed out, tests would also serve as guidance and documentation.

**Comment by LAJW** [comment]
@tautschnig Added unit tests. Hope that this proves that this code works with our codebase.

**Comment by LAJW** [comment]
I admit my trust in the library's correctness has been misplaced.

Pragma's couldn't fix the problem. Switched to a different library (Which passes on our CI without warnings): https://github.com/akrzemi1/Optional/
Added a header wrapping optional as a `optionalt` so we can easily replace it later when we gain access to C++17.

**Comment by LAJW** [comment]
Alright, I've manually fixed the library (there were 2 lines in need of changing). Re-requesting review.

File: src/util/optional.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
+/*******************************************************************\
```
Comment (thk123) [inline src/util/optional.h:1]: I'm not sure I understand the purpose of this file?

File: unit/Makefile [hunk 6]
```diff
@@ -6,6 +6,7 @@ SRC = src/expr/require_expr.cpp \
       unit_tests.cpp \
       catch_example.cpp \
       util/expr_iterator.cpp \
+      optional.cpp \
```
Comment (thk123) [inline unit/Makefile]: This should be in the util subfolder.

File: unit/optional.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ Module: nonstd::optional unit tests
+
+ Author: Diffblue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <catch.hpp>
+#include <util/optional.h>
+
+TEST_CASE("Optional without a value")
```
Comment (thk123) [inline unit/optional.cpp]: As per the [guidelines](https://github.com/diffblue/cbmc/blob/master/CODING_STANDARD.md#unit-tests) could you tag your tests appropriately:

```
[core][util][optional]
```

File: unit/optional.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+{
+  optionalt<bool> maybe_value;
+  REQUIRE(maybe_value.has_value()==false);
+  try
+  {
+    maybe_value.value();
+    REQUIRE(false); // Should have thrown an exception
+  }
+  catch(const bad_optional_accesst&) { }
```
Comment (thk123) [inline unit/optional.cpp]: Catch provides the ability to verify exceptions: https://github.com/philsquared/Catch/blob/master/docs/assertions.md#exceptions

File: src/util/optional.h (hunk 5 above)
Comment (LAJW) [reply src/util/optional.h:1]: Description updated.

**Comment by LAJW** [comment]
@thk123 This and `std::optional` doesn't allow references. `boost::optional` does.

**Comment by peterschrammel** [comment]
@LAJW, rebase please

**Pull request merged**